### PR TITLE
Call impression URL when displaying sponsored homepage shelf

### LINKED ISSUE #9720: Call impression URL when displaying sponsored homepage shelf
When implementing https://github.com/mozilla/addons-frontend/issues/9719 to fetch the sponsored shelf through the specific API we're building in https://github.com/mozilla/addons-server/issues/15617 we need to call an impression URL, that should be part of the response from the API. See also https://github.com/mozilla/addons-server/issues/15618 for the impression endpoint implementation.

One open question is around whether we call one endpoint for the whole shelf or multiple ones.

Maybe consider using https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon to make the call.

### PR DESCRIPTION
**Pull request by bobsilverberg**
~~Depends on https://github.com/mozilla/addons-frontend/pull/9758~~
Fixes #9720

### CODE REVIEW

File: config/development.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ module.exports = {
         "'self'",
         addonsServerDevCDN,
         analyticsHost,
+        apiDevHost,
```
Comment (willdurand) [inline config/development.js:35]: mm why is this needed?

File: src/amo/components/SponsoredAddonsShelf/index.js [hunk 2]
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
...
+    }
   }
 
+  sendImpressionPing = () => {
+    const { _config, _navigator, shelfData } = this.props;
+
+    if (_config.get('enableFeatureUseAdzerkForSponsoredShelf') && shelfData) {
+      const { impressionData, impressionURL } = shelfData;
+
+      if (impressionData && impressionURL) {
```
Comment (willdurand) [inline src/amo/components/SponsoredAddonsShelf/index.js:104]: Should we log a debug message (at least) when we don't have data/url?

File: src/amo/components/SponsoredAddonsShelf/index.js [hunk 3]
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
...
+        sendBeacon({
+          _navigator,
+          data: formatDataForPing({
+            data: impressionData,
+            key: 'impression_data',
+          }),
+          urlString: impressionURL,
```
Comment (willdurand) [inline src/amo/components/SponsoredAddonsShelf/index.js:111]: I suppose the API is needed in the CSP because this url is absolute. I am wondering if that has any security implications. We could potentially make this URL relative but maybe it's fine as it is currently?

File: tests/unit/amo/components/TestSponsoredAddonsShelf.js [hunk 4]
```diff
@@ -216,6 +262,82 @@ describe(__filename, () => {
 
       sinon.assert.notCalled(dispatchSpy);
     });
+
+    it('sends a ping for the impression on mount', () => {
```
Comment (willdurand) [inline tests/unit/amo/components/TestSponsoredAddonsShelf.js:266]: is it "ping" or beacon?

File: src/amo/components/SponsoredAddonsShelf/index.js [hunk 5]
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
...
+
+  componentDidUpdate(prevProps: InternalProps) {
+    const { shelfData } = this.props;
+
+    if (prevProps.shelfData !== shelfData) {
+      this.sendImpressionPing();
+    }
   }
 
+  sendImpressionPing = () => {
```
Comment (willdurand) [inline src/amo/components/SponsoredAddonsShelf/index.js:98]: is it "ping" or "beacon"?

File: tests/unit/amo/utils/test_index.js [hunk 6]
```diff
@@ -259,4 +260,48 @@ describe(__filename, () => {
...
+      sendBeacon({ _log, _navigator, urlString, data });
+      sinon.assert.calledWith(_navigator.sendBeacon, urlString, data);
+    });
+
+    it('should not send a becaon if navigator does not exist', () => {
+      const urlString = 'https://www.mozilla.org';
+      const _log = getFakeLogger();
+
+      sendBeacon({ _log, _navigator: null, urlString });
+      sinon.assert.calledWith(
```
Comment (willdurand) [inline tests/unit/amo/utils/test_index.js:290]: Asserting that `sendBeacon` isn't actually called seems a bit better than asserting a log statement. I feel like we should not have to use `_log` in the function (and only use `log`).

**Comment by bobsilverberg** [comment]
Thanks for the review, @willdurand. I am just about to land the dependency and then I will rebase this, which may require some changes. I'll reply to your comments after doing that.

File: config/development.js (hunk 1 above)
Comment (bobsilverberg) [reply config/development.js:35]: I think you answered your own question below. It's needed to enable the `sendBeacon` for the new `/shelves/sponsored/impression/` endpoint when developing locally.

File: src/amo/components/SponsoredAddonsShelf/index.js (hunk 2 above)
Comment (bobsilverberg) [reply src/amo/components/SponsoredAddonsShelf/index.js:106]: Sure, good idea.

File: src/amo/components/SponsoredAddonsShelf/index.js (hunk 3 above)
Comment (bobsilverberg) [reply src/amo/components/SponsoredAddonsShelf/index.js:113]: We actually already include the apiHost in the `connectSrc` for all other environments. The only change in this patch is that we're adding it to the `development` config, so I don't think there are any security implications.

File: tests/unit/amo/components/TestSponsoredAddonsShelf.js (hunk 4 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestSponsoredAddonsShelf.js:266]: I'm calling it a ping as a colloquial term. You're right, it's a beacon. I'll change it.

File: src/amo/components/SponsoredAddonsShelf/index.js (hunk 5 above)
Comment (bobsilverberg) [reply src/amo/components/SponsoredAddonsShelf/index.js:100]: I've changed this too.

File: tests/unit/amo/utils/test_index.js (hunk 6 above)
Comment (bobsilverberg) [reply tests/unit/amo/utils/test_index.js:290]: I would *love* to be able to assert that `sendBeacon` wasn't called, but the difficulty here is that we're testing a case where `_navigator` doesn't exist, which means that `_navigator.sendBeacon` is not a mock, and therefore there is no way for us to assert that. I agree that relying on log statements in a test is less than ideal, but this seemed like the best way to actually have a test for this, when we don't have a mock we can check.

**Comment by bobsilverberg** [comment]
Ok, @willdurand, I have rebased it and addressed the issues you raised. I accidentally squashed the changes into the initial commit, because of the rebase. I hope that isn't too much of a hassle for you.

**Review by willdurand — APPROVED** [review]
r+wc

**Comment by bobsilverberg** [comment]
Thanks @willdurand! Is this an r+wc because you wanted to make sure I saw your facepalm, or is there another comment that needs to be addressed before I can merge this?

**Comment by willdurand** [comment]
I don't think there is any other comment
